### PR TITLE
[BEAM-8379] Cache Eviction

### PR DESCRIPTION
**Pull request by KevinGG**
1. Implemented cache eviction whenever Python interpreter exits.
2. Cache for PCollections is grouped by PCollections as the
Interactive Beam user flow is now data-centric. And cache including
its eviction is  managed by a global interactive environment instance
created/retrieved/reset implicitly by runners in the same main
thread/loop.
3. Unified Python version requirement to 3.6 for interactive features
and their tests for simplicity. We don't have test suite between py35
and py36 (both exclusive) anyway, so there is no need to check by
3.5.3 when some mock test features are only available in 3.6 and later.
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by KevinGG** [comment]
Run Python PreCommit

**Comment by KevinGG** [comment]
unhashable type: 'list' [while running 'numbers assert per window/Match']

Maybe use tuples.

**Comment by KevinGG** [comment]
R: @pabloem 
Hi Pablo, this PR only adds cache eviction logic and simplified the Python version check for prerequisites and tests.

**Comment by KevinGG** [comment]
R: @davidyan74 
R: @rohdesamuel 

Thanks!

File: sdks/python/apache_beam/runners/interactive/interactive_environment.py [hunk 1]
```diff
@@ -74,10 +83,10 @@ def __init__(self, cache_manager=None):
     self._pipeline_results = {}
     # Always watch __main__ module.
     self.watch('__main__')
-    # Do a warning level logging if current python version is below 3.5.3.
-    if sys.version_info < (3, 5, 3):
+    # Do a warning level logging if current python version is below 3.6.
+    if sys.version_info < (3, 6):
       self._is_py_version_ready = False
-      logging.warning('Interactive Beam requires Python 3.5.3+.')
+      logging.warning('Interactive Beam requires Python 3.6+.')
```
Comment (rohdesamuel) [inline sdks/python/apache_beam/runners/interactive/interactive_environment.py]: Out of curiosity, why does it require 3.6+?

Comment (KevinGG) [reply sdks/python/apache_beam/runners/interactive/interactive_environment.py]: The assertions in unittest module for if something is called such as `assert_called` are not available in some older Python versions.
To make it even worse, some features introduced in Python3 has been patched back to some Python2 versions but not some older Python3 versions. So it's possible to see tests passing in Py3.6+ and Py2.7+ but fail for Py3.5.
The 3.5.3 is only a prerequisite for dependency `jsons` (not `json`) that allows us to serialize any Python object into json format without providing schema of the data. That's how we can build dataframe from arbitrary data implicitly.

Comment (pabloem) [reply sdks/python/apache_beam/runners/interactive/interactive_environment.py]: So it sounds like the interactive functionality is available in 3.5.3, but the particular assert method we use in unit tests is only available on 3.6+, right? Would it make sense for us to support 3.5.3+, and skip those tests on <3.6? If we force 3.6+ for support, we'd lose 3.5.3+ users, right?

Perhaps it's fine to do only 3.6+, but just wanting to double click on this.

**Comment by KevinGG** [comment]
> So it sounds like the interactive functionality is available in 3.5.3, but the particular assert method we use in unit tests is only available on 3.6+, right? Would it make sense for us to support 3.5.3+, and skip those tests on <3.6? If we force 3.6+ for support, we'd lose 3.5.3+ users, right?
> 
> Perhaps it's fine to do only 3.6+, but just wanting to double click on this.

Yes! Since we don't have tests for 3.5.3 but only 3.5, 3.6, 3.7, so we'll just say 3.6 in the warning and check for 3.6 in tests for simplicity.

**Pull request merged**